File: go/internal/mylib/math.go
package mylib

import (
	"fmt"
	"math/big"
	"math/bits"
)

// This code works well only on 32 or 64-bit environments.
//
//	32 (for 32-bit environment)
//	64 (for 64-bit environment)
const intSize = 32 << (^uint(0) >> 63)

func Abs(x int) int {
	// -1 (when x is negative), 0 (otherwise)
	sign := x >> (intSize - 1)

	// using two's complement
	return (x ^ sign) - sign
}

func Gcd(a, b int) int {
	x, y := uint(Abs(a)), uint(Abs(b))
	if x == 0 {
		return int(y)
	}
	if y == 0 {
		return int(x)
	}

	ntz_x := bits.TrailingZeros(x)
	ntz_y := bits.TrailingZeros(y)
	x >>= ntz_x
	y >>= ntz_y

	for {
		if x >= y {
			x -= y
			if x == 0 {
				return int(y << min(ntz_x, ntz_y))
			}
		} else {
			x, y = y-x, x
		}

		x >>= bits.TrailingZeros(x)
	}
}

func Lcm(a, b int) int {
	result := uint((Abs(a) / Gcd(a, b)) * Abs(b))
	if result&(1<<(intSize-1)) != 0 {
		panic("the result is overflow")
	}

	return int(result)
}

// If both arguments are zero, the returned value is (0, 0, 0).
// Otherwise, the returned value is (gcd(a, b), x, y) such that x and y are integers, and ax + by = gcd(a, b).
// Note: 'x' and 'y' are a pair of Bézout's coefficients for (a, b).
func ExtGcd(a, b int) (int, int, int) {
	sgn := func(x int) int {
		if x < 0 {
			return -1
		} else {
			return 1
		}
	}

	if a == 0 && b == 0 {
		return 0, 0, 0
	}

	gcd, x, y := auxExtGcd(int(Abs(a)), int(Abs(b)))
	return gcd, x * sgn(a), y * sgn(b)
}

func auxExtGcd(a, b int) (int, int, int) {
	if b == 0 {
		return int(Abs(a)), 1, 0
	}

	gcd, s, t := auxExtGcd(b, a%b)
	return gcd, t, s - a/b*t
}

func Pow(b, e int) int {
	result := 1
	for e > 0 {
		if e&1 != 0 {
			result *= b
		}
		e >>= 1
		b *= b
	}
	return result
}

func InvMod(a, m int) (int, error) {
	if m == 0 {
		return 0, fmt.Errorf("modulo must no be zero")
	}
	gcd, result, _ := ExtGcd(a, m)
	if gcd != 1 {
		return gcd, fmt.Errorf("there is no modular multiplicative inverse because GCD(%d, %d) = %d != 1", a, m, gcd)
	}

	return result, nil
}

func PowerMod(b, e, m int) int {
	if m == 0 {
		panic("modulo must not be zero")
	}

	if e < 0 {
		inv_mod, err := InvMod(b, m)
		if err != nil {
			panic("base is not invertible for the given modulus")
		}
		return auxPowerMod(inv_mod, -e, m)
	} else {
		return auxPowerMod(b, e, m)
	}
}

func auxPowerMod(b, e, m int) int {
	base := big.NewInt(int64(b))
	exp := big.NewInt(int64(e))
	modulus := big.NewInt(int64(m))

	result := new(big.Int)
	result.Exp(base, exp, modulus)

	return int(result.Int64())
}

func isqrtAux(c, n int) int {
	if c == 0 {
		return 1
	} else {
		k := (c - 1) >> 1
		a := isqrtAux(c>>1, n>>(2*k+2))
		return (a << k) + (n>>(k+2))/a
	}
}

func Isqrt(n int) int {
	if n < 0 {
		panic("argument must not be negative")
	}
	if n == 0 {
		return 0
	}

	if a := isqrtAux((bits.Len(uint(n))-1)>>1, n); n < a*a {
		return a - 1
	} else {
		return a
	}

}

// Returns the largest exponent 'e' for which pow(base, e) does not exceed 'n'.
func MaxExp(n, base int) int {
	e := 0
	for n >= base {
		n /= base
		e += 1
	}

	return e
}

func Binomial(n, k int) int {
	// C(n,k) = C(n,n-k)
	if k > n/2 {
		k = n - k
	}

	result := 1
	for i := 1; i <= k; i++ {
		result = (result * (n - k + i)) / i
	}

	return result
}

func Factorial(n int) int {
	if n < 0 {
		panic("argument must not be negative")
	}

	result := 1
	for ; n > 1; n-- {
		result *= n
	}

	return result
}

func PowBig(base *big.Int, e int) *big.Int {
	exp := big.NewInt(int64(e))
	result := new(big.Int)

	return result.Exp(base, exp, nil)
}

func IsTriangular(n int) bool {
	tmp := 8*n + 1
	tmp_sqrt := Isqrt(tmp)

	return tmp_sqrt&1 == 1 && tmp_sqrt*tmp_sqrt == tmp
}

func IsSquare(n int) bool {
	n_sqrt := Isqrt(n)

	return n_sqrt*n_sqrt == n
}

func IsPentagonal(n int) bool {
	tmp := 24*n + 1
	tmp_sqrt := Isqrt(tmp)

	return tmp_sqrt%6 == 5 && tmp_sqrt*tmp_sqrt == tmp
}

func IsHexagonal(n int) bool {
	tmp := 8*n + 1
	tmp_sqrt := Isqrt(tmp)

	return tmp_sqrt%4 == 3 && tmp_sqrt*tmp_sqrt == tmp
}

File: haskell/src/Mylib/Math.hs
module Mylib.Math (
    extGcd,
    invMod,
    powerMod,
    powerModExn,
    isqrt,
    kronecker,
    binomial,
    factorial,
    maxPower,
    numOfDigits,
    getCTZ,
    isPalindrome,
    isPandigital,
    isPandigitalNZ,
    isTriangular,
    isSquare,
    isPentagonal,
    isHexagonal,
) where

import Data.Bits (shiftL, (.|.))
import Data.Maybe (fromMaybe)

import Mylib.Util (bitLength)

extGcd :: Integral int => int -> int -> (int, int, int)
extGcd 0 0 = (0, 0, 0)
extGcd a b =
    let (g, x, y) = extGcd' (abs a) (abs b)
     in (g, sgn a * x, sgn b * y)
  where
    sgn x = if x < 0 then -1 else 1
    extGcd' x 0 = (abs x, 1, 0)
    extGcd' x y =
        let (g, s, t) = extGcd' y (mod x y)
         in (g, t, s - div x y * t)
{-# SPECIALIZE extGcd :: Int -> Int -> (Int, Int, Int) #-}
{-# SPECIALIZE extGcd :: Integer -> Integer -> (Integer, Integer, Integer) #-}

invMod :: Integral int => int -> int -> Maybe int
invMod _ 0 = Nothing -- modulo must not be zero
invMod a m
    | g /= 1 = Nothing -- there is no modular multiplicative inverse
    | otherwise = Just result
  where
    (g, result, _) = extGcd a m
{-# SPECIALIZE invMod :: Int -> Int -> Maybe Int #-}
{-# SPECIALIZE invMod :: Integer -> Integer -> Maybe Integer #-}

powerMod :: Int -> Int -> Int -> Maybe Int
powerMod _ _ 0 = Nothing -- "modulo must not be zero"
powerMod b e m
    | e < 0 = do
        case invMod b m of
            Just x ->
                Just
                    (fromIntegral (powerMod' (fromIntegral x) (fromIntegral (-e)) (fromIntegral m) 1))
            Nothing -> Nothing
    | otherwise =
        Just
            (fromIntegral (powerMod' (fromIntegral b) (fromIntegral e) (fromIntegral m) 1))

powerMod' :: Integer -> Integer -> Integer -> Integer -> Integer
powerMod' b e m result
    | e == 0 = result
    | odd e = powerMod' ((b * b) `mod` m) (e `div` 2) m ((result * b) `mod` m)
    | otherwise = powerMod' ((b * b) `mod` m) (e `div` 2) m result

powerModExn :: Int -> Int -> Int -> Int
powerModExn b e m = fromMaybe (error "") (powerMod b e m)

isqrt :: Integral int => int -> int
isqrt n
    | n == 0 = 0
    | n < 0 = error "argument must not be negative"
    | otherwise =
        let a = isqrt' (div (bitLength n - 1) 2) n
         in if n < a * a then a - 1 else a
  where
    isqrt' 0 _ = 1
    isqrt' c n' =
        let k = div (c - 1) 2
            a = isqrt' (div c 2) (div n' (2 ^ (2 * k + 2)))
         in a * 2 ^ k + div (div n' (2 ^ (k + 2))) a
{-# SPECIALIZE isqrt :: Int -> Int #-}
{-# SPECIALIZE isqrt :: Integer -> Integer #-}

kronecker :: Integral int => int -> int -> int
kronecker a 0 = if abs a == 1 then 1 else 0
kronecker a n
    | n < 0 =
        if a < 0 then kronecker' (-1) a (negate n) else kronecker' 1 a (negate n)
    | otherwise = kronecker' 1 a n
{-# SPECIALIZE kronecker :: Int -> Int -> Int #-}
{-# SPECIALIZE kronecker :: Integer -> Integer -> Integer #-}

kronecker' :: Integral int => int -> int -> int -> int
kronecker' sign a n
    | even n =
        if even a
            then
                0
            else
                let (new_n, new_sign) = auxKronecker sign n a
                 in kronecker'' new_sign (mod a new_n) new_n
    | otherwise = kronecker'' sign (mod a n) n
{-# SPECIALIZE kronecker' :: Int -> Int -> Int -> Int #-}
{-# SPECIALIZE kronecker' :: Integer -> Integer -> Integer -> Integer #-}

kronecker'' :: Integral int => int -> int -> int -> int
kronecker'' sign 0 1 = sign
kronecker'' _ 0 _ = 0
kronecker'' sign a n =
    kronecker'' new_sign (mod n new_a) new_a
  where
    (new_a, tmp_sign) = auxKronecker sign a n
    new_sign = if mod new_a 4 == 3 && mod n 4 == 3 then negate tmp_sign else tmp_sign
{-# SPECIALIZE kronecker'' :: Int -> Int -> Int -> Int #-}
{-# SPECIALIZE kronecker'' :: Integer -> Integer -> Integer -> Integer #-}

auxKronecker :: Integral int => int -> int -> int -> (int, int)
auxKronecker sign x y
    | odd x = (x, sign)
    | otherwise = (x `div` (2 ^ ntz_x), new_sign)
  where
    ntz_x = getCTZ x
    new_sign =
        if even ntz_x
            then
                sign
            else
                let tmp = y `mod` 8
                 in if tmp == 3 || tmp == 5 then (-sign) else sign
{- FOURMOLU_DISABLE -}
{-# SPECIALIZE auxKronecker :: Int -> Int -> Int -> (Int, Int) #-}
{-# SPECIALIZE auxKronecker :: Integer -> Integer -> Integer -> (Integer, Integer) #-}
{- FOURMOLU_ENABLE -}

binomial :: Integral int => int -> int -> int
binomial _ 0 = 1
binomial n k
    | n == k = 1
    | n < k = 0
    | (div n 2) < k = binomial n (n - k)
    | otherwise = foldl (\acc i -> acc * (n - k + i) `div` i) 1 [1 .. k]
{-# SPECIALIZE binomial :: Int -> Int -> Int #-}
{-# SPECIALIZE binomial :: Integer -> Integer -> Integer #-}

factorial :: Integral int => int -> int
factorial 0 = 1
factorial n = n * factorial (n - 1)
{-# SPECIALIZE factorial :: Int -> Int #-}
{-# SPECIALIZE factorial :: Integer -> Integer #-}

maxPower :: Integral int => int -> int -> Int
maxPower n base =
    aux n 0
  where
    aux x e
        | x >= base = aux (div x base) (succ e)
        | otherwise = e
{-# SPECIALIZE maxPower :: Int -> Int -> Int #-}
{-# SPECIALIZE maxPower :: Integer -> Integer -> Int #-}

numOfDigits :: Integral int => int -> int -> Int
numOfDigits n base = (maxPower n base) + 1
{-# SPECIALIZE numOfDigits :: Int -> Int -> Int #-}
{-# SPECIALIZE numOfDigits :: Integer -> Integer -> Int #-}

getCTZ :: Integral int => int -> Int
getCTZ n = getCTZ' n 0
  where
    getCTZ' x cnt = if odd x then cnt else getCTZ' (x `div` 2) (succ cnt)
{-# SPECIALIZE getCTZ :: Int -> Int #-}
{-# SPECIALIZE getCTZ :: Integer -> Int #-}

isPalindrome :: Integral int => int -> int -> Bool
isPalindrome n base = (isPalindrome' n 0) == n
  where
    isPalindrome' x acc =
        if x == 0
            then
                acc
            else
                isPalindrome' (x `div` base) (acc * base + (x `mod` base))
{-# SPECIALIZE isPalindrome :: Int -> Int -> Bool #-}
{-# SPECIALIZE isPalindrome :: Integer -> Integer -> Bool #-}

isPandigital :: Int -> Bool
isPandigital n =
    (makeBits n 0) == (shiftL 1 ((maxPower n 10) + 1)) - 1
  where
    makeBits :: Int -> Int -> Int
    makeBits 0 bits = bits
    makeBits i bits = makeBits (div i 10) ((.|.) bits (shiftL 1 (mod i 10)))

isPandigitalNZ :: Int -> Bool
isPandigitalNZ n =
    checkZero n && isPandigital (n * 10)
  where
    checkZero :: Int -> Bool
    checkZero 0 = True
    checkZero i
        | (mod i 10) == 0 = False
        | otherwise = checkZero (div i 10)

isTriangular :: Integral int => int -> Bool
isTriangular n =
    mod tmp_sqrt 2 == 1 && tmp_sqrt * tmp_sqrt == tmp
  where
    tmp = 8 * n + 1
    tmp_sqrt = isqrt tmp
{-# SPECIALIZE isTriangular :: Int -> Bool #-}
{-# SPECIALIZE isTriangular :: Integer -> Bool #-}

isSquare :: Integral int => int -> Bool
isSquare n =
    n_sqrt * n_sqrt == n
  where
    n_sqrt = isqrt n
{-# SPECIALIZE isSquare :: Int -> Bool #-}
{-# SPECIALIZE isSquare :: Integer -> Bool #-}

isPentagonal :: Integral int => int -> Bool
isPentagonal n =
    mod tmp_sqrt 6 == 5 && tmp_sqrt * tmp_sqrt == tmp
  where
    tmp = 24 * n + 1
    tmp_sqrt = isqrt tmp
{-# SPECIALIZE isPentagonal :: Int -> Bool #-}
{-# SPECIALIZE isPentagonal :: Integer -> Bool #-}

isHexagonal :: Integral int => int -> Bool
isHexagonal n =
    mod tmp_sqrt 4 == 3 && tmp_sqrt * tmp_sqrt == tmp
  where
    tmp = 8 * n + 1
    tmp_sqrt = isqrt tmp
{-# SPECIALIZE isHexagonal :: Int -> Bool #-}
{-# SPECIALIZE isHexagonal :: Integer -> Bool #-}
